File: scripts/test-sponsor-framework-v7-53-2.py
#!/usr/bin/env python3
"""Regression checks for the WPI 7.53.2 sponsor framework."""
from __future__ import annotations

import json
import re
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
RELEASE = "7.53.2"


def fail(message: str) -> None:
    raise SystemExit(f"Sponsor framework check failed: {message}")


config_path = ROOT / "data/sponsors/index.json"
runtime_path = ROOT / "data/sponsors/runtime.js"
framework_path = ROOT / "js/sponsor-framework-v7-53-2.js"
css_path = ROOT / "css/sponsor-framework-v7-53-2.css"

for path in (config_path, runtime_path, framework_path, css_path, ROOT / "docs/SPONSOR_FRAMEWORK.md"):
    if not path.exists():
        fail(f"missing {path.relative_to(ROOT)}")

config = json.loads(config_path.read_text())
if config.get("release") != RELEASE:
    fail(f"config release is {config.get('release')!r}, expected {RELEASE}")
if config.get("status") != "ready_no_active_campaigns":
    fail("release must ship with no active advertiser")
if config.get("campaigns") != []:
    fail("campaign list must be empty until a real sponsor is approved")
if config.get("privacy") != {
    "collectsPersonalData": False,
    "usesCookies": False,
    "usesLocalStorage": False,
    "clickTrackingEnabled": False,
    "outboundAttribution": "UTM parameters only",
}:
    fail("privacy contract changed")
if "never influences WPI rankings" not in config.get("disclosure", ""):
    fail("independence disclosure is missing")

placements = config.get("placements") or []
ids = [placement.get("id") for placement in placements]
expected = {
    "club.presenting", "club.inline", "club.region",
    "team.presenting", "team.inline",
    "rankings.presenting", "rankings.inline",
    "tournaments.presenting", "tournaments.inline",
    "regions.directory",
}
if set(ids) != expected or len(ids) != len(set(ids)):
    fail(f"placement inventory changed: {ids}")
for placement in placements:
    if not placement.get("label") or not placement.get("format"):
        fail(f"placement metadata incomplete: {placement.get('id')}")
    if not placement.get("mount", {}).get("selectors"):
        fail(f"placement mount missing: {placement.get('id')}")

runtime = runtime_path.read_text().strip()
match = re.fullmatch(r"window\.WPI_SPONSOR_CONFIG = (.*);", runtime, flags=re.S)
if not match:
    fail("runtime assignment format is invalid")
if json.loads(match.group(1)) != config:
    fail("runtime.js is not synchronized with index.json")

framework = framework_path.read_text()
for token in (
    'rel="sponsored noopener noreferrer"',
    'utm_source',
    'utm_medium',
    'wpi:sponsors-rendered',
    'campaign.status === "active"',
    'document.querySelectorAll("[data-wpi-sponsor-placement]")',
):
    if token not in framework:
        fail(f"framework contract missing {token}")
for forbidden in ("localStorage", "sessionStorage", "document.cookie", "fetch(", "XMLHttpRequest"):
    if forbidden in framework:
        fail(f"privacy regression: framework contains {forbidden}")

root_pages = [
    "club.html", "team.html", "rankings.html", "clubs.html", "tournaments.html", "quicksilver-cup-2026.html",
    "12u-boys.html", "12u-girls.html", "14u-boys.html", "14u-girls.html",
    "16u-boys.html", "16u-girls.html", "18u-boys.html", "18u-girls.html",
]
sub_pages = [
    "tournaments/boys-superfinals/index.html",
    "tournaments/girls-club-championships/index.html",
    "tournaments/jo-boys/index.html",
    "tournaments/jo-girls/index.html",
    "tournaments/quicksilver-cup/index.html",
]
for rel in root_pages:
    text = (ROOT / rel).read_text()
    for ref in (
        'css/sponsor-framework-v7-53-2.css?v=7.53.2',
        'data/sponsors/runtime.js?v=7.53.2',
        'js/sponsor-framework-v7-53-2.js?v=7.53.2',
    ):
        if ref not in text:
            fail(f"{rel} missing {ref}")
for rel in sub_pages:
    text = (ROOT / rel).read_text()
    for ref in (
        '../../css/sponsor-framework-v7-53-2.css?v=7.53.2',
        '../../data/sponsors/runtime.js?v=7.53.2',
        '../../js/sponsor-framework-v7-53-2.js?v=7.53.2',
    ):
        if ref not in text:
            fail(f"{rel} missing {ref}")

club_js = (ROOT / "js/club-profile-v7-53-0.js").read_text()
if "renderSponsorModule" in club_js or 'id="club-partners"' in club_js or "WPI partner opportunity" in club_js:
    fail("legacy always-visible club sponsor solicitation remains")
if "rankings and results are not influenced by sponsorship" not in club_js:
    fail("club profile independence language was removed")

# Protect the positive data baseline while adding commercial infrastructure.
data_lines = (ROOT / "data.js").read_text().splitlines()
def assigned_json(prefix: str):
    line = next((line for line in data_lines if line.startswith(prefix)), None)
    if line is None or not line.endswith(";"):
        fail(f"could not read {prefix}")
    return json.loads(line[len(prefix):-1])
rankings = assigned_json("window.CPI_RANKINGS = ")
clubs = assigned_json("window.CPI_CLUBS = ")
if len(rankings) != 724:
    fail(f"ranking count changed: {len(rankings)}")
if len(clubs) != 182:
    fail(f"club count changed: {len(clubs)}")
if sum(bool(club.get("website")) for club in clubs) != 177:
    fail("club website coverage changed")

jo_results = json.loads((ROOT / "data/tournaments/jo-results-2026.json").read_text())
if jo_results.get("summary", {}).get("teamPlacements") != 976:
    fail("JO placement count changed")

print("WPI 7.53.2 sponsor framework check passed.")
print(f"  {len(placements)} placements; 0 active campaigns; {len(rankings)} rankings; {len(clubs)} clubs; 976 JO placements")

File: scripts/test-sponsor-framework-v7-53-2.js
#!/usr/bin/env node
"use strict";

const fs = require("fs");
const vm = require("vm");
const path = require("path");
const root = path.resolve(__dirname, "..");
const code = fs.readFileSync(path.join(root, "js/sponsor-framework-v7-53-2.js"), "utf8");

function run(campaigns) {
  const inserted = [];
  const mount = { insertAdjacentElement(position, node) { inserted.push({ position, node }); } };
  class Element {
    constructor(tag) { this.tagName = tag; this.dataset = {}; this.attributes = {}; this.className = ""; this.innerHTML = ""; }
    setAttribute(name, value) { this.attributes[name] = value; }
    remove() { this.removed = true; }
  }
  const classes = new Set();
  const document = {
    currentScript: { src: "https://littlezip12.github.io/CPI/js/sponsor-framework-v7-53-2.js?v=7.53.4" },
    readyState: "complete",
    body: { classList: { contains() { return false; } } },
    documentElement: { classList: { add(value) { classes.add(value); }, remove(value) { classes.delete(value); } } },
    createElement(tag) { return new Element(tag); },
    querySelector(selector) {
      if (selector === ".rankings-hero") return mount;
      return null;
    },
    querySelectorAll() { return []; },
    addEventListener() {},
    dispatchEvent() {},
  };
  const config = {
    release: "7.53.2",
    disclosure: "Sponsorship never influences WPI rankings, results, placements, or editorial decisions.",
    outboundAttribution: { source: "waterpoloindex", medium: "sponsor", campaignParameter: "utm_campaign", placementParameter: "utm_content" },
    placements: [{ id: "rankings.presenting", pageTypes: ["rankings"], label: "Rankings Presented by", format: "presenting", mount: { selectors: [".rankings-hero"], position: "beforebegin" } }],
    campaigns,
  };
  const window = {
    WPI_SPONSOR_CONFIG: config,
    location: { pathname: "/WPI/rankings.html", search: "", href: "https://littlezip12.github.io/CPI/rankings.html" },
    setTimeout(fn) { fn(); return 1; },
    clearTimeout() {},
  };
  const sandbox = { window, document, URL, URLSearchParams, Date, console, CustomEvent: class CustomEvent { constructor(type, init) { this.type = type; this.detail = init?.detail; } } };
  vm.runInNewContext(code, sandbox, { filename: "sponsor-framework-v7-53-2.js" });
  return { inserted, classes, api: window.WPI_SPONSORS };
}

const empty = run([]);
if (empty.inserted.length !== 0) throw new Error("Empty campaign inventory rendered page space");
if (!empty.api || empty.api.activeCampaignCount() !== 0) throw new Error("Sponsor API did not initialize safely");

const active = run([{
  id: "approved-test",
  name: "Approved Test Partner",
  slug: "approved-test-partner",
  status: "active",
  placements: ["rankings.presenting"],
  startDate: "2026-01-01",
  endDate: "2026-12-31",
  website: "https://example.com/support",
  message: "Supporting youth water polo.",
  cta: "Visit partner",
  priority: 100,
}]);
if (active.inserted.length !== 1) throw new Error(`Expected one sponsor card, found ${active.inserted.length}`);
const html = active.inserted[0].node.innerHTML;
for (const token of [
  "Approved Test Partner",
  "Rankings Presented by",
  "rel=\"sponsored noopener noreferrer\"",
  "utm_source=waterpoloindex",
  "utm_medium=sponsor",
  "utm_campaign=approved-test-partner",
  "utm_content=rankings.presenting",
  "never influences WPI rankings",
]) {
  if (!html.includes(token)) throw new Error(`Rendered sponsor card missing ${token}`);
}
if (!active.classes.has("wpi-has-active-sponsors")) throw new Error("Active sponsor state was not set");
console.log("WPI 7.53.2 sponsor browser runtime check passed.");
